scipy/stats/_qmc.py

```diff
@@ -481,6 +485,186 @@ def _perturb_discrepancy(sample: np.ndarray, i1: int, i2: int, k: int,
+        eps: float = 1e-10
+) -> np.ndarray:
+    """Randomly jitter `sample` until they are unique.
+
+    If the samples are not unique, the number of regions in our
+    tessellation will be less than the number of input samples.
+    """
+    vals, count = np.unique(sample, return_counts=True)
+    if np.any(vals[count > 1]):
+        offset = rng.uniform(-eps, eps, size=(len(sample), sample.shape[1]))
+        sample = sample * (1 + offset)
+    return sample
+
+
+def _lloyd_centroidal_voronoi_tessellation(
+        sample: np.ndarray,
+        decay: float,
+        rng: Union[np.random.Generator, np.random.RandomState]
+) -> np.ndarray:
+    """Lloyd's algorithm iteration.
+
+    Based on the implementation of Stéfan van der Walt:
+
+    https://github.com/stefanv/lloyd
+
+    which is:
+
+        Copyright (c) 2021-04-21 Stéfan van der Walt
+        https://github.com/stefanv/lloyd
+        MIT License
+
+    Parameters
+    ----------
+    sample : array_like (n, d)
+        The sample to iterate on.
+    decay : float
+        Number of iterations. Too many iterations tend to cluster the sample
+        as a hypersphere.
+    rng : `numpy.random.Generator`
+        Random number generator.
+
+    Returns
+    -------
+    sample : array_like (n, d)
+        The sample after an iteration of Lloyd's algorithm.
+
+    """
+    sample = _jitter_sample(sample, rng=rng)
+    new_sample = np.empty_like(sample)
+
+    voronoi = spatial.Voronoi(sample)
+
+    for ii, idx in enumerate(voronoi.point_region):
+        # the region is a series of indices into self.voronoi.vertices
+        # remove point at infinity, designated by index -1
+        region = [i for i in voronoi.regions[idx] if i != -1]
+
+        # get the vertices for this region
+        verts = voronoi.vertices[region]
+
+        # clipping would be wrong, we need to intersect
+        # verts = np.clip(verts, 0, 1)
+
+        # move sample towards centroids:
+        # Centroid in n-D is the mean for uniformly distributed nodes
+        # of a geometry.
+        centroid = np.mean(verts, axis=0)
+        new_sample[ii] = sample[ii] + (centroid - sample[ii]) * decay
+
+    # only update sample to centroid within the region
+    is_valid = np.all(np.logical_and(new_sample >= 0, new_sample <= 1), axis=1)
+    sample[is_valid] = new_sample[is_valid]
+
+    return sample
+
+
+def lloyd_centroidal_voronoi_tessellation(
+        sample: npt.ArrayLike,
+        *,
+        n_iters: IntNumber = 10,
+        seed: SeedType = None
+) -> np.ndarray:
+    """Centroidal Voronoi Tessellation.
+
+    Perturb a sample of points in :math:`[0, 1]^d` using Lloyd's algorithm.
+
+    Parameters
+    ----------
+    sample : array_like (n, d)
+        The sample to iterate on. With ``n`` the number of sample and ``d``
+        the dimension.
+    n_iters : int, optional
+        Number of iterations. Too many iterations tend to cluster the sample
+        as a hypersphere. Default is 10.
```

Comment:
A convergence parameter could be good. But `minimized global energy may be interpreted as the L1-norm of the error achieved when the function x2 is interpolated on the mesh domain using a piecewise linear function which is linear in each cell of the Voronoi diagram of the mesh vertices` sounds a bit weird. I would just go with the L1-norm of the sample itself. Basically the metric from the test.
```python
        def l1_norm(sample):
            l1 = cdist(sample, sample, 'cityblock')
            return np.min(l1[l1.nonzero()])
```

Reply:
> an actual time limit 

I'm a fan of a wall-clock time limit, but are there any other SciPy functions that have this sort of thing? I was under the impression that it would be discouraged. In any case, it is a lot easier to add than remove, so I'd suggest 

> a number of iterations like you specify here, but they instead default to 0

If this should be the default behavior, I think the meaning of `maxiter=np.inf` would be more obvious than `maxiter=0. But should this be the default behavior, @tupui? What do you think about relying on the convergence criterion rather than `maxiter` by default?

> a convergence parameter

Looks like a convergence parameter has been added.